workflow/scripts/osemosys_global/powerplant/user_defined_capacity.py

```diff
@@ -0,0 +1,255 @@
+        max_cap_techs_df.infer_objects().fillna(0,
+                  inplace=True)
+        
+        max_cap_techs_df = max_cap_techs_df[['REGION',
+                                             'TECHNOLOGY',
+                                             'YEAR',
+                                             'VALUE']]
+        
+        # Append existing TotalAnnualMaxCapacityInvestment data with MAX_BUILD
+        df_max_cap_inv = pd.concat([df_max_cap_invest, max_cap_techs_df]).drop_duplicates()
+
+        # Print TotalAnnualMaxCapacityInvestment.csv with MAX_BUILD
+        df_max_cap_inv.drop_duplicates(subset=['REGION', 
+                                               'TECHNOLOGY',
+                                               'YEAR'],
+                                       keep='last',
+                                       inplace=True)
+
+        # For technologies with start year before model start year, add to ResidualCapacity
+        df_res_cap_ud = df_min_cap_inv.copy().loc[df_min_cap_inv.copy()['YEAR'] < min(get_years(start_year, end_year))]
+        df_res_cap_ud.rename(columns={'YEAR':'start_year'},
+                             inplace=True)
+        df_res_cap_ud_final = pd.DataFrame(list(itertools.product(df_res_cap_ud['TECHNOLOGY'].unique(),
+                                                                  get_years(start_year, end_year))
+                                                ),
+                                           columns = ['TECHNOLOGY',
+                                                      'YEAR']
+                                           )
+        df_res_cap_ud_final = pd.merge(df_res_cap_ud_final,
+                                       df_res_cap_ud,
+                                       how='left',
+                                       on=['TECHNOLOGY'])
+        df_res_cap_ud_final['TECH'] = df_res_cap_ud_final['TECHNOLOGY'].str[3:6]
+        df_res_cap_ud_final['OP_LIFE'] = df_res_cap_ud_final['TECH'].map(op_life_dict)
+        df_res_cap_ud_final['end_year'] = (df_res_cap_ud_final['OP_LIFE'] 
+                                           + df_res_cap_ud_final['start_year'])
+        df_res_cap_ud_final = df_res_cap_ud_final.loc[df_res_cap_ud_final['YEAR'] 
+                                                      >= df_res_cap_ud_final['start_year']]
+        df_res_cap_ud_final = df_res_cap_ud_final.loc[df_res_cap_ud_final['YEAR'] 
+                                                      <= df_res_cap_ud_final['end_year']]
+        df_res_cap_ud_final['REGION'] = region_name
+        df_res_cap_ud_final = df_res_cap_ud_final[['REGION',
+                                                   'TECHNOLOGY',
+                                                   'YEAR',
+                                                   'VALUE']]
+
+        df_res_cap = pd.concat([df_res_cap, df_res_cap_ud_final 
+                                if not df_res_cap_ud_final.empty else None])
```

Comment:
This needs a groupby, but has been addressed in the other PR thread in [this commit ](https://github.com/OSeMOSYS/osemosys_global/pull/206/commits/ea6f89aea8a4bfbe53f87e403fc205f03a732a1b)